activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,227 @@
+        mattr_accessor :backtrace_cleaner, default: ActiveSupport::BacktraceCleaner.new
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+
+        class << self
+          @@backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
+
+          # Updates the context used to construct the query log tags.
+          # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def update(**options)
+            context.merge!(**options.symbolize_keys)
+            self.cached_comment = nil
+          end
+
+          def tags_available?
+            components.present? || inline_annotations.present?
+          end
+
+          # Returns an SQL comment +String+ containing the query log tags.
+          # Sets and returns a cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def comment
+            if cache_query_log_tags
+              self.cached_comment ||= uncached_comment
+            else
+              uncached_comment
+            end
+          end
+
+          def uncached_comment
+            content = tag_content
+            if content.present?
+              "/*#{escape_sql_comment(content)}*/"
+            end
+          end
+
+          # Returns a +String+ containing any inline comments from +with_annotation+.
+          def inline_comment
+            return nil unless inline_annotations.present?
+            "/*#{escape_sql_comment(inline_tag_content)}*/"
+          end
+
+          # Manually clear the comment cache.
+          def clear_comment_cache!
+            self.cached_comment = nil
+          end
+
+          # Annotate any query within `&block`. Can be nested.
+          def with_annotation(comment, &block)
+            self.inline_annotations.push(comment)
+            block.call if block.present?
+          ensure
+            self.inline_annotations.pop
+          end
+
+          # Return the set of active inline annotations from +with_annotation+.
+          def inline_annotations
+            context[:inline_annotations] ||= []
+          end
+
+          # QueryLogTags +component+ methods
+
+          # Set during Rails boot in lib/active_record/railtie.rb
+          def application # :nodoc:
+            context[:application_name]
+          end
+
+          def pid # :nodoc:
+            Process.pid
+          end
+
+          def connection_config # :nodoc:
+            ActiveRecord::Base.connection_db_config
+          end
+
+          def socket # :nodoc:
+            connection_config.socket
+          end
+
+          def db_host # :nodoc:
+            connection_config.host
+          end
+
+          def database # :nodoc:
+            connection_config.database
+          end
+
+          def line # :nodoc:
+            # if we're ever calling this component we should disable caching
+            self.cache_query_log_tags = false
```

Comment:
Do you also need to call `clear_comment_cache!` here?

Reply:
I don't think so, because once it's disabled the cache [isn't referenced at all](https://github.com/rails/rails/pull/42240/files#diff-a32ffa60fc5a3f3cd4356aa08e2457a381ae22a58ff5deae74df98db249aec71R124).

Reply:
Some more context - I don't think the tag components will change in run time in most cases, so this is really just catching the edge case where someone in local development may get caught off guard by the cache when they elect to use this component.

Would love to hear from more users to confirm this assumption though.

Reply:
Resolving as `line` has been removed.